### DISCUSSION TITLE
[Not started] Add the Matter 1.0 Device Types and Application Clusters supported by Google Home Ecosystem?

### DISCUSSION THREAD

**Discussion by Hedda** [mfucci/node-matter#133] [Ideas]
### Google Home Ecosystem interoperable compatibility for its Matter 1.0 supported device types and actions?

Please consider prioritizing support for Matter 1.0 Device Types and Application Clusters now supported by Google Home Ecosystem:

https://developers.home.google.com/matter/get-started

https://developers.home.google.com/matter/supported-devices

https://console.home.google.com/

Google Home Ecosystem includes controlling devices via Google Home app, Google Assistant, Google Nest speakers and Google Hub.

https://en.wikipedia.org/wiki/Google_Nest_(smart_speakers)

Use case # 1 : [Google Home App](https://support.google.com/chromecast/answer/7071794) **opt-in Public Preview Program app pre-release** is now rolling out updates with initial Matter support so prioritizing that would mean users will be able to use Google Home app and Google Assistant as a client to control supported device types.

Use case # 2 : [Google Nest Hub (2nd gen)](https://en.wikipedia.org/wiki/Google_Nest_(smart_speakers)#Second_generation) **opt-in Developer Preview Program firmware** will soon roll out updates with initial Matter 1.0 controller/server support so would mean that [Google Nest Hub (2nd gen)](https://en.wikipedia.org/wiki/Google_Nest_(smart_speakers)#Second_generation) will be able to act as a device bridge for non-Matter devices.

https://developers.home.google.com/matter/release-notes

https://support.google.com/googlenest/answer/6343937

https://support.google.com/googlenest/answer/12494697

FYI, an opt-in Developer Preview Program sign-up page already exists that you can access on your Android or iPhone (iOS) phone/tablet by following this local URL It will allow you to request an invite when the program is available -> `googlehome://settings/app-preview-program`

https://www.makeuseof.com/sign-up-for-public-preview-google-home-app/

You can also now enable opt-in Developer Preview Program firmware seperate per Google Nest speaker or display and Google Chromecast:

https://www.makeuseof.com/join-google-home-preview-program/

Other user interfaces including soon-to-be-launched opt-in preview version of Google Home with a new look and initial Matter 1.0 support:

- Google Home website (and beta web interface)
  *  https://home.google.com/welcome/
- Google Home (for Android) in Google Play Store
  *  https://play.google.com/store/apps/details?id=com.google.android.apps.chromecast.app
- Google Home (for iOS) in the Apple App Store
  * https://apps.apple.com/se/app/google-home/id680819774

Here is an overview of Google's developers documentation for Matter Device Type conversion to Google Home Ecosystem Type:

Matter Device Type | Application Cluster(s) | Matter Spec | Home Ecosystem Type
-- | -- | -- | -- 
Contact Sensor | Boolean State | 1.0 | Sensor
Dimmable Light | Level Control | 1.0 | Light
Door Lock | Door Lock | 1.0 | Lock
Extended Color Light | Color Control (MoveToHueAndSaturation) | 1.0 | Light
Humidity Sensor | Relative Humidity Measurement | 1.0 | Sensor
Illuminance Sensor | Illuminance Measurement | 1.0 | Sensor
On/Off Light | On/Off | 1.0 | Light
Occupancy Sensor | Occupancy Sensing | 1.0 | Sensor
Plug | Level ControlOn/Off | 1.0 | Outlet
Speaker | Level ControlOn/Off | 1.0 | Speaker
Temperature Sensor | Temperature Measurement | 1.0 | Thermostat
Thermostat | Thermostat | 1.0 | Thermostat
Window Covering | Window Covering | 1.0 | Blinds

Google's developer center allows filtering of supported use cases by supported device integration and that now includes "Matter":

https://developers.home.google.com/use-cases

So far these are the initial device categories and device types that Google now looks to have added support for in Google Home Ecosystem:

- Lighting
  * Light -> https://developers.home.google.com/use-cases?device=light
  * Outlet -> https://developers.home.google.com/use-cases?device=outlet
  * Switch -> https://developers.home.google.com/use-cases?device=switch

- Climate
  * Thermostat -> https://developers.home.google.com/use-cases?device=thermostat
  * Air conditioner
  * Blinds -> https://developers.home.google.com/use-cases?device=blinds
  * Fan -> https://developers.home.google.com/use-cases?device=fan

- Safety & security
  * Lock -> https://developers.home.google.com/use-cases?device=lock
  * Sensor -> https://developers.home.google.com/use-cases?device=sensor

- Media & Technology
  * Speaker -> https://developers.home.google.com/use-cases?device=speaker

PS: Obviously Matter 1.0 covers many more additional Device Types and Application Clusters so I assume that Google developers will surely prioritize extending support to additional use cases to at least support all use cases that have a matching Google Home Ecosystem Type:

Ref: https://github.com/project-chip/connectedhomeip/tree/v1.0-branch/src/app/clusters (Note! `1.0-branch` only for spec compatibility)

**Comment by mfucci** [comment]
Thanks for this info and the list!

I think the 1st step here is to ensure that node-matter is compatible Google Home (as you can see in the bugs, it doesn't work fully with iOs yet).
Do you have the beta Google Home installed on your phone and do you think you can do a quick test?

Do you see more node-matter acting a device in the eco-system (so it can be controlled by Google Home) or as a "controller" (it can be added on the fabric by Google Home and you can use node-matter to write your own automation scripts)?
Obviously we want to support both, but we will need to focus on which use case to focus on first.

**Reply by Hedda** [reply]
> Do you see more node-matter acting a device in the eco-system (so it can be controlled by Google Home) or as a "controller" (it can be added on the fabric by Google Home and you can use node-matter to write your own automation scripts)?
> Obviously we want to support both, but we will need to focus on which use case to focus on first.

I believe your former use case scenario with node-matter acting as a server to allow devices from Node-RED to be used in the Google Home ecosystem will probably be the most common use case scenario today (e.i. enable users to trigger devices via voice control on their Google Nest smart speaker and smart display products or from the Google mobile app), but looking farther ahead at their roadmaps I guess that might change in the future as there are definitely also very valid future use cases for the reverse as well because Google Home already have support for loads of non-Matter devices so if they will allow and enable Google Nest Hub (2nd gen) to act as a Matter bridge for many of those non-Matter devices.

Personally I currently only use Google Home (as Google Assistant) solely as a client platform because we have either a Google Nest speaker or a Google Nest Hub smart display in almost each and every room in our house as primarily a voice user interface, and the user interface of the Google Home mobile app is actually also very user friendly as a frontend for managing those devices that need to have voice control and/or manual mobile app for remote control as an option. That is, we have none of our automation running in the Google Home ecosystem itself, (and as far as I know, the Google Home ecosystem did not really support running automations on its own before).

https://en.wikipedia.org/wiki/Google_Nest_(smart_speakers) 

With the upcoming Google Home app (Preview Program release) it will be possible to use Google Nest Hub (2nd gen) as a "controller", meaning that you can add your devices to your Google Home ecosystem with it acting as a server that can share the devices with other platforms via the Matter 1.0 standard. If that works as well as it sounds on paper then that means a use case is that users will probably be able add devices that do not support Matter to the Google Home ecosystem and it can then act as a Matter bridge, e.g. share the devices with node-matter acting as a client.

https://www.androidpolice.com/big-google-home-app-update-coming/

FYI, Google also announced that upcoming Google Home app update will bring an Script Editor feature for advanced routines:

https://9to5google.com/2022/10/04/google-home-web-app/

So I also understand that the upcoming Google Home app from Google's Preview Program(s) they will even add nativly integrated features to be able to create and run automations locally on a [Google Nest Hub (2nd gen)](https://en.wikipedia.org/wiki/Google_Nest_(smart_speakers)#Second_generation), with the idea that you just need to get your devices into the Google Home ecosystem via the Matter 1.0 standard and then you can control and automate everything from there, (which I think will be nice for new users as the first entry into home automation). 

https://www.androidpolice.com/google-home-device-triggers-assistant-early/

> Do you have the beta Google Home installed on your phone and do you think you can do a quick test?

Sorry, I don't have a working test environment right now (and my main setup is in "production" due to stability for the whole family and WAF). Hope that listing use cases is at least enough to help get you started on adding device types that Google Home support.

**Reply by Hedda** [reply]
By the way, please note that I think all of these above-mentioned use cases will soon also apply to the Amazon Alexa ecosystem and Amazon Echo or Amazon Tap platform, (which I personally do not use), as I know many other users use Amazon Echo or Amazon Tap smart speakers and Amazon Echo Show in more or less the exact same use case scenarios for Amazon Alexa voice control. E.g. presumably today primarily with Amazon Echo and Amazon Echo Show as client platforms for Amazon Alexa voice control, but in the future possible also as a controller and Matter 1.0 bridge for non-Matter devices already supported by the Amazon Echo platform.

https://en.wikipedia.org/wiki/Amazon_Alexa

https://en.wikipedia.org/wiki/Amazon_Echo

https://en.wikipedia.org/wiki/Amazon_Echo_Show

**Reply by Apollon77** [reply]
Yes this is all a part of "the issue" rught now ... all stuff is" on the horizon" but not there - beside with Apple iOS 16.1/16.2 (upcoming). Amazon devices "till end of the year", but then most likely only with an Android App this year (iOS next year). Googe - as you wrote - also "coming when nest hub and app is out". This all makes it veery complex and difficult to do interoperability testing.

Thats why we right now foxus getting the needed minimum clusters ready for a "minimal" device as first - including bridged devices (needed for ioBroker and also zigbee2mqtt case). After that - and when it works with iOS because only "really available reference" - then we can see if someone can test  with other ecosystems ... then we see if we need to further adjust things (hopefully not) :-)

There is still a long way to go

**Comment by Hedda** [comment]
Oh, by the way, I believe that a related early adopter use case for node-matter will probably be to act as an MQTT-to-Matter bridge for users of the popular Zigbee2MQTT project in order to bridge existing already-owned Zigbee devices to other platforms, like the Google Home / Google Nest (and Amazon Alexa /  Amazon Echo) ecosystems, which primarily will be Matter clients as they do not support all the many different Zigbee devices that the Z2M project already support today (which are loads of various Zigbee based products in different device type categories):

 https://www.zigbee2mqtt.io/

https://github.com/Koenkk/zigbee2mqtt

https://www.zigbee2mqtt.io/supported-devices/

And maybe another related but more more complexed use case scenario as a Zigbee-to-Matter bridge for advanced users Node-RED that instead choose to directly use the zigbee-herdsman (which is the Node.js Zigbee library that the Z2M project above uses as a dependency):

https://github.com/Koenkk/zigbee-herdsman

https://github.com/Koenkk/zigbee-herdsman-converters

**Reply by Apollon77** [reply]
Yes the developer already is "looking at us" library wise, but also here a lot of stuff needs to be done.

**Comment by Hedda** [comment]
Looks like you can also (optionally?) build your own Android app for commissioning physical and virtual Matter devices to Google Home(?):

https://developers.home.google.com/samples/matter-app

https://developers.home.google.com/codelabs/matter-sample-app#0

https://developers.home.google.com/codelabs/matter-device-virtual

https://developers.home.google.com/samples/matter-app

https://developers.home.google.com/matter/tools/virtual-device-controller

https://github.com/google-home/sample-app-for-matter-android/releases

https://developers.home.google.com/matter/tools/android-studio

https://developers.home.google.com/matter/apis/home
